src/core/hard_forks.zig

```diff
@@ -0,0 +1,85 @@
+const std = @import("std");
+const ArrayList = std.ArrayList;
+const allocator = std.mem.Allocator;
+const Slot = @import("slot.zig").Slot;
+pub const HardFork = struct { slot: Slot, count: usize };
+pub const HardForks = struct {
+    hard_forks: ArrayList(HardFork),
+
+    const Self = @This();
+
+    pub fn default(alloc: allocator) Self {
+        return .{ .hard_forks = ArrayList(HardFork).init(alloc) };
+    }
+
+    pub fn register(self: *Self, new_slot: Slot) !void {
+        const maybe_index = for (self.hard_forks.items, 0..) |hard_fork, index| {
+            if (hard_fork.slot == new_slot) break index;
+        } else null;
+
+        if (maybe_index) |index| {
+            self.hard_forks.items[index] = .{ .slot = new_slot, .count = self.hard_forks.items[index].count +| 1 };
+        } else {
+            try self.hard_forks.append(.{ .slot = new_slot, .count = 1 });
+        }
+        std.mem.sort(HardFork, self.hard_forks.items, {}, struct {
+            pub fn lessThan(_: void, a: HardFork, b: HardFork) bool {
+                return a.slot < b.slot;
+            }
+        }.lessThan);
+    }
+
+    pub fn deinit(self: *Self) void {
+        self.hard_forks.deinit();
+    }
+
+    pub fn get_forks(self: *const Self) []HardFork {
+        return self.hard_forks.items;
+    }
+
+    pub fn get_hash_data(self: *Self, slot: Slot, parent_slot: Slot) ?[8]u8 {
```

Comment:
can we change this to `*const Self` - all fcns which dont modify the ptr should be const